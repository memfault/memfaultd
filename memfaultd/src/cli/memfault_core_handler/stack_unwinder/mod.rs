//
// Copyright (c) Memfault, Inc.
// See License.txt for details
mod eh_frame_finder;
mod stacktrace_format;
mod unwind_handler;
mod unwinder;

pub use eh_frame_finder::EhFrameFinderImpl;
pub use unwind_handler::UnwindHandler;
pub use unwinder::{UnwindFrameContext, Unwinder};
