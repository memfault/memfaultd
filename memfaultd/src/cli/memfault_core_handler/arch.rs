//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::{collections::HashMap, mem::size_of};

use super::stack_unwinder::UnwindFrameContext;
use crate::cli::memfault_core_handler::elf;

use cfg_if::cfg_if;
use gimli::{Register, Vendor};

cfg_if! {
    if #[cfg(target_arch = "aarch64")] {
        pub use libc::user_regs_struct as ElfGRegSet;
        use gimli::AArch64;
        pub fn get_stack_pointer(regs: &ElfGRegSet) -> usize {
            regs.sp as usize
        }
        pub fn get_program_counter(regs: &ElfGRegSet) -> usize {
            regs.pc as usize
        }
        pub fn get_return_register(regs: &HashMap<Register, usize>) -> Option<usize> {
            // Register X30 is the LR on AArch64
            regs.get(&AArch64::X30).copied()
        }
        pub const fn return_register_idx() -> Register {
            AArch64::X30
        }
        pub fn set_stack_pointer(regs: &mut HashMap<Register, usize>, sp: usize)  {
            regs.insert(AArch64::SP, sp);
        }
        pub use elf::header::EM_AARCH64 as ELF_TARGET_MACHINE;
        pub use elf::header::ELFCLASS64 as ELF_TARGET_CLASS;
        impl From<&ElfGRegSet> for UnwindFrameContext {
            fn from(regs: &ElfGRegSet) -> Self {
                let mut ctx = UnwindFrameContext::default();
                ctx.registers.insert(AArch64::X0, regs.regs[0] as usize);
                ctx.registers.insert(AArch64::X1, regs.regs[1] as usize);
                ctx.registers.insert(AArch64::X2, regs.regs[2] as usize);
                ctx.registers.insert(AArch64::X3, regs.regs[3] as usize);
                ctx.registers.insert(AArch64::X4, regs.regs[4] as usize);
                ctx.registers.insert(AArch64::X5, regs.regs[5] as usize);
                ctx.registers.insert(AArch64::X6, regs.regs[6] as usize);
                ctx.registers.insert(AArch64::X7, regs.regs[7] as usize);
                ctx.registers.insert(AArch64::X8, regs.regs[8] as usize);
                ctx.registers.insert(AArch64::X9, regs.regs[9] as usize);
                ctx.registers.insert(AArch64::X10, regs.regs[10] as usize);
                ctx.registers.insert(AArch64::X11, regs.regs[11] as usize);
                ctx.registers.insert(AArch64::X12, regs.regs[12] as usize);
                ctx.registers.insert(AArch64::X13, regs.regs[13] as usize);
                ctx.registers.insert(AArch64::X14, regs.regs[14] as usize);
                ctx.registers.insert(AArch64::X15, regs.regs[15] as usize);
                ctx.registers.insert(AArch64::X16, regs.regs[16] as usize);
                ctx.registers.insert(AArch64::X17, regs.regs[17] as usize);
                ctx.registers.insert(AArch64::X18, regs.regs[18] as usize);
                ctx.registers.insert(AArch64::X19, regs.regs[19] as usize);
                ctx.registers.insert(AArch64::X20, regs.regs[20] as usize);
                ctx.registers.insert(AArch64::X21, regs.regs[21] as usize);
                ctx.registers.insert(AArch64::X22, regs.regs[22] as usize);
                ctx.registers.insert(AArch64::X23, regs.regs[23] as usize);
                ctx.registers.insert(AArch64::X24, regs.regs[24] as usize);
                ctx.registers.insert(AArch64::X25, regs.regs[25] as usize);
                ctx.registers.insert(AArch64::X26, regs.regs[26] as usize);
                ctx.registers.insert(AArch64::X27, regs.regs[27] as usize);
                ctx.registers.insert(AArch64::X28, regs.regs[28] as usize);
                ctx.registers.insert(AArch64::X29, regs.regs[29] as usize);
                ctx.registers.insert(AArch64::X30, regs.regs[30] as usize);
                ctx.registers.insert(AArch64::SP, regs.sp as usize);
                ctx.registers.insert(AArch64::PC, regs.pc as usize);

                ctx
            }
        }
    } else if #[cfg(target_arch = "x86_64")] {
        pub use libc::user_regs_struct as ElfGRegSet;
        use gimli::X86_64;
        pub fn get_stack_pointer(regs: &ElfGRegSet) -> usize {
            regs.rsp as usize
        }
        pub fn get_program_counter(regs: &ElfGRegSet) -> usize {
            regs.rip as usize
        }
        pub fn get_return_register(regs: &HashMap<Register, usize>) -> Option<usize> {
            regs.get(&X86_64::RA).copied()
        }
        pub const fn return_register_idx() -> Register {
            X86_64::RA
        }
        pub fn set_stack_pointer(regs: &mut HashMap<Register, usize>, sp: usize)  {
            regs.insert(X86_64::RSP, sp);
        }
        pub use elf::header::EM_X86_64 as ELF_TARGET_MACHINE;
        pub use elf::header::ELFCLASS64 as ELF_TARGET_CLASS;

        impl From<&ElfGRegSet> for UnwindFrameContext {
            fn from(regs: &ElfGRegSet) -> Self {
                let mut ctx = UnwindFrameContext::default();
                ctx.registers.insert(X86_64::RAX, regs.rax as usize);
                ctx.registers.insert(X86_64::RBX, regs.rbx as usize);
                ctx.registers.insert(X86_64::RCX, regs.rcx as usize);
                ctx.registers.insert(X86_64::RDX, regs.rdx as usize);
                ctx.registers.insert(X86_64::RSI, regs.rsi as usize);
                ctx.registers.insert(X86_64::RDI, regs.rdi as usize);
                ctx.registers.insert(X86_64::RBP, regs.rbp as usize);
                ctx.registers.insert(X86_64::RSP, regs.rsp as usize);
                ctx.registers.insert(X86_64::R8, regs.r8 as usize);
                ctx.registers.insert(X86_64::R9, regs.r9 as usize);
                ctx.registers.insert(X86_64::R10, regs.r10 as usize);
                ctx.registers.insert(X86_64::R11, regs.r11 as usize);
                ctx.registers.insert(X86_64::R12, regs.r12 as usize);
                ctx.registers.insert(X86_64::R13, regs.r13 as usize);
                ctx.registers.insert(X86_64::R14, regs.r14 as usize);
                ctx.registers.insert(X86_64::R15, regs.r15 as usize);
                ctx.registers.insert(X86_64::RFLAGS, regs.eflags as usize);
                ctx.registers.insert(X86_64::CS, regs.cs as usize);
                ctx.registers.insert(X86_64::FS, regs.fs as usize);
                ctx.registers.insert(X86_64::GS, regs.gs as usize);
                ctx.registers.insert(X86_64::SS, regs.ss as usize);
                ctx.registers.insert(X86_64::DS, regs.ds as usize);
                ctx.registers.insert(X86_64::ES, regs.es as usize);
                ctx.registers.insert(X86_64::FS_BASE, regs.fs_base as usize);
                ctx.registers.insert(X86_64::GS_BASE, regs.gs_base as usize);

                ctx
            }
        }
    } else if #[cfg(target_arch = "arm")] {
        pub use libc::user_regs as ElfGRegSet;
        pub fn get_stack_pointer(regs: &ElfGRegSet) -> usize {
            regs.arm_sp as usize
        }
        pub fn get_program_counter(regs: &ElfGRegSet) -> usize {
            regs.arm_pc as usize
        }
        pub fn set_stack_pointer(_regs: &mut HashMap<Register, usize>, sp: usize)  {
            todo!()
        }
        pub fn get_return_register(_regs: &HashMap<Register, usize>) -> Option<usize> {
            todo!()
        }
        pub const fn return_register_idx() -> Register {
            todo!()
        }
        pub use elf::header::EM_ARM as ELF_TARGET_MACHINE;
        pub use elf::header::ELFCLASS32 as ELF_TARGET_CLASS;

        impl From<&ElfGRegSet> for UnwindFrameContext {
            fn from(regs: &ElfGRegSet) -> Self {
                todo!()
            }
        }
    } else if #[cfg(target_arch = "x86")] {
        pub use libc::user_regs_struct as ElfGRegSet;
        pub fn get_stack_pointer(regs: &ElfGRegSet) -> usize {
            regs.esp as usize
        }
        pub fn get_program_counter(regs: &ElfGRegSet) -> usize {
            regs.eip as usize
        }
        pub fn set_stack_pointer(_regs: &mut HashMap<Register, usize>, sp: usize)  {
            todo!()
        }
        pub fn get_return_register(_regs: &HashMap<Register, usize>) -> Option<usize> {
            todo!()
        }
        pub const fn return_register_idx() -> Register {
            todo!()
        }
        pub use elf::header::EM_386 as ELF_TARGET_MACHINE;
        pub use elf::header::ELFCLASS32 as ELF_TARGET_CLASS;

        impl From<&ElfGRegSet> for UnwindFrameContext {
            fn from(regs: &ElfGRegSet) -> Self {
                todo!()
            }
        }
    }
    else {
        // Provide dummy symbols for unsupported architectures. This is preferable to
        // a compile time error, as we want to be able to compile memfaultd for all
        // architectures, but we don't need register access for all of them. Currently
        // these registers are only used to filter out stack memory from coredumps.
        pub struct ElfGRegSet;
        pub fn get_stack_pointer(_regs: &ElfGRegSet) -> usize {
            0
        }
        pub fn get_program_counter(_regs: &ElfGRegSet) -> usize {
            0
        }
        pub fn set_stack_pointer(_regs: &mut HashMap<Register, usize>, sp: usize)  {
            todo!()
        }
        pub fn get_return_register(_regs: &HashMap<Register, usize>) -> Option<usize> {
            None
        }
        pub const fn return_register_idx() -> Register {
            todo!()
        }
        impl From<&ElfGRegSet> for UnwindFrameContext {
            fn from(regs: &ElfGRegSet) -> Self {
                UnwindFrameContext::default()
            }
        }
    }
}

// Function definitions for coredump thread filter support. If the target architecture
// is not supported, these functions will always return false.
cfg_if! {
    if #[cfg(any(
        target_arch = "aarch64",
        target_arch = "arm",
        target_arch = "x86",
        target_arch = "x86_64"
    ))] {
        pub const fn coredump_thread_filter_supported() -> bool {
            true
        }
    } else {
        pub const fn coredump_thread_filter_supported() -> bool {
            false
        }
    }
}

cfg_if! {
    if #[cfg(target_arch = "aarch64")] {
        pub fn get_vendor() -> Vendor {
            Vendor::AArch64
        }
    } else {
        pub fn get_vendor() -> Vendor {
            Vendor::Default
        }
    }
}

// Function definitions for stacktrace support. If the target architecture is not
// supported, these functions will always return false.
cfg_if! {
    if #[cfg(any(
        target_arch = "aarch64",
        target_arch = "x86_64"
    ))] {
        pub const fn stacktrace_supported() -> bool {
            true
        }
    } else {
        pub const fn stacktrace_supported() -> bool {
            false
        }
    }
}

#[cfg(target_endian = "little")]
pub use elf::header::ELFDATA2LSB as ELF_TARGET_ENDIANNESS;

#[cfg(target_endian = "big")]
pub use elf::header::ELFDATA2MSB as ELF_TARGET_ENDIANNESS;

pub const WORD_SIZE: usize = size_of::<usize>();
