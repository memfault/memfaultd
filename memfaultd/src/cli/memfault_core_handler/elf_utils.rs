//
// Copyright (c) Memfault, Inc.
// See License.txt for details
//! Utilities for reading ELF files

use std::{
    collections::HashMap,
    io::{Read, Seek, SeekFrom},
};

use super::elf::section_header::SectionHeader;
use super::{
    core_elf_note::iterate_elf_notes,
    elf::header::{Header, EI_CLASS, EI_DATA, ELFCLASS, ELFMAG, EV_CURRENT, SELFMAG, SIZEOF_EHDR},
};
use super::{
    core_elf_note::ElfNote,
    elf::program_header::{ProgramHeader, SIZEOF_PHDR},
};
use crate::cli::memfault_core_handler::arch::{ELF_TARGET_ENDIANNESS, ELF_TARGET_MACHINE};

use eyre::{eyre, Result};
use libc::{PF_X, PT_NOTE};
use scroll::Pread;

/// Read the ELF header from an input stream.
///
/// Will return an error if the header is invalid, or the stream is too short.
pub fn read_elf_header<R: Read>(input_stream: &mut R) -> Result<Header> {
    let mut header_buf = [0u8; SIZEOF_EHDR];
    input_stream.read_exact(&mut header_buf)?;

    let elf_header = *Header::from_bytes(&header_buf);
    if !verify_elf_header(&elf_header) {
        return Err(eyre!("Invalid ELF header"));
    }

    Ok(elf_header)
}

/// Get the base execution address of the provided ELF file.
///
/// The same as `get_base_addr_from_reader`, but reads the program headers from the input stream.
pub fn get_base_addr_from_reader<R: Read + Seek>(
    input_stream: &mut R,
    elf_header: &Header,
) -> Result<usize> {
    input_stream.seek(SeekFrom::Start(elf_header.e_phoff as _))?;
    let program_headers = read_program_headers(input_stream, elf_header.e_phnum as usize)?;

    get_base_address_from_program_headers(&program_headers)
}

/// Reads `count` ELF program headers from the provided input stream.
pub fn read_program_headers<R: Read>(
    input_stream: &mut R,
    count: usize,
) -> Result<Vec<ProgramHeader>> {
    let size = count * SIZEOF_PHDR;
    let mut buffer = vec![0; size];
    input_stream.read_exact(&mut buffer)?;
    Ok(ProgramHeader::from_bytes(&buffer, count))
}

pub fn read_elf_notes<R: Read + Seek>(
    input_stream: &mut R,
    program_headers: &[ProgramHeader],
) -> Result<Vec<Vec<u8>>> {
    let raw_notes = program_headers
        .iter()
        .filter(|ph| ph.p_type == PT_NOTE)
        .map(|ph| {
            let mut note_data = vec![0u8; ph.p_filesz as usize];
            input_stream.seek(SeekFrom::Start(ph.p_offset as _))?;
            input_stream.read_exact(&mut note_data)?;

            Ok(note_data)
        })
        .collect::<Result<Vec<_>>>()?;

    Ok(raw_notes)
}

/// Convert raw note data into a list of parsed ELF notes.
pub fn parse_elf_notes(raw_notes: &[Vec<u8>]) -> Result<Vec<ElfNote<'_>>> {
    let parsed_notes = raw_notes
        .iter()
        .flat_map(|data| iterate_elf_notes(data))
        .collect::<Vec<_>>();

    Ok(parsed_notes)
}

pub fn get_build_id<R: Read + Seek>(input_stream: &mut R, elf_header: &Header) -> Result<Vec<u8>> {
    input_stream.seek(SeekFrom::Start(elf_header.e_phoff as _))?;
    let program_headers = read_program_headers(input_stream, elf_header.e_phnum as usize)?;

    let raw_notes = read_elf_notes(input_stream, &program_headers)?;
    let parsed_notes = parse_elf_notes(&raw_notes)?;

    let build_id = parsed_notes
        .iter()
        .find_map(|note| match note {
            ElfNote::GnuBuildId(build_id) => Some(build_id.to_vec()),
            _ => None,
        })
        .ok_or_else(|| eyre!("Failed to find GNU Build ID note"))?;

    Ok(build_id)
}

/// Finds the base execution address from the provided program headers.
///
/// This is the address of the first executable segment. This address is used to determine the
/// offset of the program when ASLR is enabled.
pub fn get_base_address_from_program_headers(program_headers: &[ProgramHeader]) -> Result<usize> {
    program_headers
        .iter()
        .find_map(|ph| {
            if ph.p_flags & PF_X != 0 {
                Some(ph.p_vaddr as usize)
            } else {
                None
            }
        })
        .ok_or_else(|| eyre!("No executable segment found"))
}

#[derive(Debug, Clone)]
/// ELF section representation.
///
/// Contains the section header and the section data.
pub struct Section {
    pub header: SectionHeader,
    pub data: Vec<u8>,
}

/// ELF section map utility.
///
/// Provides the ability to read sections from an ELF file by name. The sections are streamed
/// whenever requested, so we do not have to load the entire file into memory. Considering
/// that the ELF contains the text, data, and bss sections, this avoids basically doubling
/// the RAM footprint of any program loading its own elf.
pub struct SectionMap<R: Read + Seek> {
    elf_reader: R,
    name_map: HashMap<String, SectionHeader>,
}

impl<R> SectionMap<R>
where
    R: Read + Seek,
{
    // Create a new section map from an ELF file and its header.
    pub fn new(mut elf_reader: R, elf_header: &Header) -> Result<Self> {
        let strtab_idx = elf_header.e_shstrndx as usize;

        let section_headers = read_section_headers(&mut elf_reader, elf_header)?;
        // Read the section header string table
        let strtab = read_section(&mut elf_reader, section_headers[strtab_idx])?;

        // Create a map of section names to section headers
        let name_map = section_headers
            .into_iter()
            .map(|sh| {
                let name = section_name_from_strtab(&strtab, sh.sh_name as usize)?;
                Ok((name, sh))
            })
            .collect::<Result<HashMap<String, SectionHeader>>>()?;

        Ok(Self {
            elf_reader,
            name_map,
        })
    }

    /// Get a section by name.
    ///
    /// This allows us to only load sections into memory that we're actually interested in
    /// and avoid the rest.
    pub fn get_section(&mut self, name: &str) -> Result<Option<Section>> {
        let header = self.name_map.get(name);
        match header {
            None => Ok(None),
            Some(header) => {
                let section = read_section(&mut self.elf_reader, *header)?;

                Ok(Some(section))
            }
        }
    }
}

/// Given a section and an offset, return the section name.
fn section_name_from_strtab(strtab: &Section, offset: usize) -> Result<String> {
    // All strings are NULL terminated, so we can just search for the next NULL byte.
    let end = strtab.data[offset..]
        .iter()
        .position(|&c| c == 0)
        .unwrap_or(strtab.data.len() - offset);
    let name = std::str::from_utf8(&strtab.data[offset..(offset + end)])?.to_string();

    Ok(name)
}

/// Read a section from an input stream.
fn read_section<R: Read + Seek>(input_stream: &mut R, header: SectionHeader) -> Result<Section> {
    let data_size = header.sh_size as usize;

    let mut data = vec![0u8; data_size];
    input_stream.seek(SeekFrom::Start(header.sh_offset as _))?;
    input_stream.read_exact(&mut data)?;

    Ok(Section { header, data })
}

/// Read the section headers from an input stream.
fn read_section_headers<R: Read + Seek>(
    input_stream: &mut R,
    header: &Header,
) -> Result<Vec<SectionHeader>> {
    let sh_offset = header.e_shoff as _;
    let sh_count = header.e_shnum as usize;
    let sh_size = header.e_shentsize as usize;
    input_stream.seek(SeekFrom::Start(sh_offset))?;

    let mut section_header_buf = vec![0u8; sh_size];
    let mut section_headers = Vec::with_capacity(sh_count);
    for _ in 0..sh_count {
        let mut offset = 0;
        input_stream.read_exact(&mut section_header_buf)?;
        let sh = section_header_buf.gread::<SectionHeader>(&mut offset)?;

        section_headers.push(sh);
    }

    Ok(section_headers)
}

fn verify_elf_header(header: &Header) -> bool {
    &header.e_ident[0..SELFMAG] == ELFMAG
        && header.e_ident[EI_CLASS] == ELFCLASS
        && header.e_ident[EI_DATA] == ELF_TARGET_ENDIANNESS
        && header.e_version == EV_CURRENT as u32
        && header.e_ehsize == SIZEOF_EHDR as u16
        && header.e_phentsize == SIZEOF_PHDR as u16
        && header.e_machine == ELF_TARGET_MACHINE
}

#[cfg(test)]
mod test {
    use std::{fs::File, path::PathBuf};

    use super::*;

    use crate::cli::memfault_core_handler::arch::ELF_TARGET_CLASS;
    use crate::cli::memfault_core_handler::elf::header::{ELFCLASSNONE, ELFDATANONE, EM_NONE};
    use crate::cli::memfault_core_handler::test_utils::build_test_header;

    use insta::assert_debug_snapshot;
    use itertools::Itertools;
    use rstest::rstest;

    #[rstest]
    // Mismatching class (32 vs 64 bit):
    #[case(ELFCLASSNONE, ELF_TARGET_ENDIANNESS, ELF_TARGET_MACHINE)]
    // Mismatching endianness:
    #[case(ELF_TARGET_CLASS, ELFDATANONE, ELF_TARGET_MACHINE)]
    // Mismatching machine:
    #[case(ELF_TARGET_CLASS, ELF_TARGET_ENDIANNESS, EM_NONE)]
    fn test_verify_elf_header_fails_for_mismatching_arch(
        #[case] class: u8,
        #[case] endianness: u8,
        #[case] machine: u16,
    ) {
        let elf_header = build_test_header(class, endianness, machine);
        assert!(!verify_elf_header(&elf_header));
    }

    #[rstest]
    fn test_section_header_read() {
        let path = bin_path();
        let mut file = File::open(path).unwrap();

        let header = read_elf_header(&mut file).unwrap();
        let section_map = read_section_headers(&mut file, &header).unwrap();

        assert_debug_snapshot!(section_map);
    }

    #[rstest]
    fn test_section_name_read() {
        let path = bin_path();
        let mut file = File::open(path).unwrap();

        let header = read_elf_header(&mut file).unwrap();
        let section_map = SectionMap::new(&mut file, &header).unwrap();
        let names = section_map
            .name_map
            .keys()
            .map(|s| s.as_str())
            .sorted()
            .collect::<Vec<&str>>();

        assert_debug_snapshot!(names);
    }

    #[rstest]
    #[case(".text", true)]
    #[case(".data", true)]
    #[case(".bss", true)]
    #[case(".garbage", false)]
    fn test_get_section(#[case] name: &str, #[case] exists: bool) {
        let path = bin_path();
        let mut file = File::open(path).unwrap();

        let header = read_elf_header(&mut file).unwrap();
        let mut section_map = SectionMap::new(&mut file, &header).unwrap();

        let section = section_map.get_section(name).unwrap();
        assert_eq!(section.is_some(), exists);
    }

    #[test]
    fn test_get_build_id() {
        let path = bin_path();
        let mut file = File::open(path).unwrap();

        let header = read_elf_header(&mut file).unwrap();
        let build_id = get_build_id(&mut file, &header).unwrap();

        assert_eq!(build_id.len(), 20);
        assert_debug_snapshot!(build_id);
    }

    fn bin_path() -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("src/cli/memfault_core_handler/fixtures/simple_executable/simple_exe.elf")
    }
}
