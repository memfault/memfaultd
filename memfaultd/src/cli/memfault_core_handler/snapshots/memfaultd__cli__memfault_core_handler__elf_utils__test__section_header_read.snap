---
source: memfaultd/src/cli/memfault_core_handler/elf_utils.rs
expression: section_map
---
[
    SectionHeader {
        sh_name: 0,
        sh_type: "SHT_NULL",
        sh_flags: 0x0,
        sh_addr: 0x0,
        sh_offset: 0x0,
        sh_size: 0x0,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x0,
        sh_entsize: 0x0,
    },
    SectionHeader {
        sh_name: 11,
        sh_type: "SHT_PROGBITS",
        sh_flags: 0x2,
        sh_addr: 0x318,
        sh_offset: 0x318,
        sh_size: 0x1c,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x1,
        sh_entsize: 0x0,
    },
    SectionHeader {
        sh_name: 19,
        sh_type: "SHT_NOTE",
        sh_flags: 0x2,
        sh_addr: 0x338,
        sh_offset: 0x338,
        sh_size: 0x30,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x8,
        sh_entsize: 0x0,
    },
    SectionHeader {
        sh_name: 38,
        sh_type: "SHT_NOTE",
        sh_flags: 0x2,
        sh_addr: 0x368,
        sh_offset: 0x368,
        sh_size: 0x24,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x4,
        sh_entsize: 0x0,
    },
    SectionHeader {
        sh_name: 57,
        sh_type: "SHT_NOTE",
        sh_flags: 0x2,
        sh_addr: 0x38c,
        sh_offset: 0x38c,
        sh_size: 0x20,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x4,
        sh_entsize: 0x0,
    },
    SectionHeader {
        sh_name: 71,
        sh_type: "SHT_GNU_HASH",
        sh_flags: 0x2,
        sh_addr: 0x3b0,
        sh_offset: 0x3b0,
        sh_size: 0x24,
        sh_link: 0x6,
        sh_info: 0x0,
        sh_addralign: 0x8,
        sh_entsize: 0x0,
    },
    SectionHeader {
        sh_name: 81,
        sh_type: "SHT_DYNSYM",
        sh_flags: 0x2,
        sh_addr: 0x3d8,
        sh_offset: 0x3d8,
        sh_size: 0xa8,
        sh_link: 0x7,
        sh_info: 0x1,
        sh_addralign: 0x8,
        sh_entsize: 0x18,
    },
    SectionHeader {
        sh_name: 89,
        sh_type: "SHT_STRTAB",
        sh_flags: 0x2,
        sh_addr: 0x480,
        sh_offset: 0x480,
        sh_size: 0x8d,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x1,
        sh_entsize: 0x0,
    },
    SectionHeader {
        sh_name: 97,
        sh_type: "SHT_GNU_VERSYM",
        sh_flags: 0x2,
        sh_addr: 0x50e,
        sh_offset: 0x50e,
        sh_size: 0xe,
        sh_link: 0x6,
        sh_info: 0x0,
        sh_addralign: 0x2,
        sh_entsize: 0x2,
    },
    SectionHeader {
        sh_name: 110,
        sh_type: "SHT_GNU_VERNEED",
        sh_flags: 0x2,
        sh_addr: 0x520,
        sh_offset: 0x520,
        sh_size: 0x30,
        sh_link: 0x7,
        sh_info: 0x1,
        sh_addralign: 0x8,
        sh_entsize: 0x0,
    },
    SectionHeader {
        sh_name: 125,
        sh_type: "SHT_RELA",
        sh_flags: 0x2,
        sh_addr: 0x550,
        sh_offset: 0x550,
        sh_size: 0xc0,
        sh_link: 0x6,
        sh_info: 0x0,
        sh_addralign: 0x8,
        sh_entsize: 0x18,
    },
    SectionHeader {
        sh_name: 135,
        sh_type: "SHT_RELA",
        sh_flags: 0x42,
        sh_addr: 0x610,
        sh_offset: 0x610,
        sh_size: 0x18,
        sh_link: 0x6,
        sh_info: 0x18,
        sh_addralign: 0x8,
        sh_entsize: 0x18,
    },
    SectionHeader {
        sh_name: 145,
        sh_type: "SHT_PROGBITS",
        sh_flags: 0x6,
        sh_addr: 0x1000,
        sh_offset: 0x1000,
        sh_size: 0x1b,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x4,
        sh_entsize: 0x0,
    },
    SectionHeader {
        sh_name: 140,
        sh_type: "SHT_PROGBITS",
        sh_flags: 0x6,
        sh_addr: 0x1020,
        sh_offset: 0x1020,
        sh_size: 0x20,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x10,
        sh_entsize: 0x10,
    },
    SectionHeader {
        sh_name: 151,
        sh_type: "SHT_PROGBITS",
        sh_flags: 0x6,
        sh_addr: 0x1040,
        sh_offset: 0x1040,
        sh_size: 0x10,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x10,
        sh_entsize: 0x10,
    },
    SectionHeader {
        sh_name: 160,
        sh_type: "SHT_PROGBITS",
        sh_flags: 0x6,
        sh_addr: 0x1050,
        sh_offset: 0x1050,
        sh_size: 0x10,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x10,
        sh_entsize: 0x10,
    },
    SectionHeader {
        sh_name: 169,
        sh_type: "SHT_PROGBITS",
        sh_flags: 0x6,
        sh_addr: 0x1060,
        sh_offset: 0x1060,
        sh_size: 0x19c,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x10,
        sh_entsize: 0x0,
    },
    SectionHeader {
        sh_name: 175,
        sh_type: "SHT_PROGBITS",
        sh_flags: 0x6,
        sh_addr: 0x11fc,
        sh_offset: 0x11fc,
        sh_size: 0xd,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x4,
        sh_entsize: 0x0,
    },
    SectionHeader {
        sh_name: 181,
        sh_type: "SHT_PROGBITS",
        sh_flags: 0x2,
        sh_addr: 0x2000,
        sh_offset: 0x2000,
        sh_size: 0x60,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x8,
        sh_entsize: 0x0,
    },
    SectionHeader {
        sh_name: 189,
        sh_type: "SHT_PROGBITS",
        sh_flags: 0x2,
        sh_addr: 0x2060,
        sh_offset: 0x2060,
        sh_size: 0x4c,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x4,
        sh_entsize: 0x0,
    },
    SectionHeader {
        sh_name: 203,
        sh_type: "SHT_PROGBITS",
        sh_flags: 0x2,
        sh_addr: 0x20b0,
        sh_offset: 0x20b0,
        sh_size: 0x10c,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x8,
        sh_entsize: 0x0,
    },
    SectionHeader {
        sh_name: 213,
        sh_type: "SHT_INIT_ARRAY",
        sh_flags: 0x3,
        sh_addr: 0x3db8,
        sh_offset: 0x2db8,
        sh_size: 0x8,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x8,
        sh_entsize: 0x8,
    },
    SectionHeader {
        sh_name: 225,
        sh_type: "SHT_FINI_ARRAY",
        sh_flags: 0x3,
        sh_addr: 0x3dc0,
        sh_offset: 0x2dc0,
        sh_size: 0x8,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x8,
        sh_entsize: 0x8,
    },
    SectionHeader {
        sh_name: 237,
        sh_type: "SHT_DYNAMIC",
        sh_flags: 0x3,
        sh_addr: 0x3dc8,
        sh_offset: 0x2dc8,
        sh_size: 0x1f0,
        sh_link: 0x7,
        sh_info: 0x0,
        sh_addralign: 0x8,
        sh_entsize: 0x10,
    },
    SectionHeader {
        sh_name: 155,
        sh_type: "SHT_PROGBITS",
        sh_flags: 0x3,
        sh_addr: 0x3fb8,
        sh_offset: 0x2fb8,
        sh_size: 0x48,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x8,
        sh_entsize: 0x8,
    },
    SectionHeader {
        sh_name: 246,
        sh_type: "SHT_PROGBITS",
        sh_flags: 0x3,
        sh_addr: 0x4000,
        sh_offset: 0x3000,
        sh_size: 0x10,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x8,
        sh_entsize: 0x0,
    },
    SectionHeader {
        sh_name: 252,
        sh_type: "SHT_NOBITS",
        sh_flags: 0x3,
        sh_addr: 0x4010,
        sh_offset: 0x3010,
        sh_size: 0x8,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x1,
        sh_entsize: 0x0,
    },
    SectionHeader {
        sh_name: 257,
        sh_type: "SHT_PROGBITS",
        sh_flags: 0x30,
        sh_addr: 0x0,
        sh_offset: 0x3010,
        sh_size: 0x2b,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x1,
        sh_entsize: 0x1,
    },
    SectionHeader {
        sh_name: 1,
        sh_type: "SHT_STRTAB",
        sh_flags: 0x0,
        sh_addr: 0x0,
        sh_offset: 0x303b,
        sh_size: 0x10a,
        sh_link: 0x0,
        sh_info: 0x0,
        sh_addralign: 0x1,
        sh_entsize: 0x0,
    },
]
